urlaa051fbd3d_WheezePuppet_collegeDiversity_tgz-pJ8-edu_umw_cpsc_collegediversity_SimJ8/src/edu/umw/cpsc/collegediversity/UpperclassHousingSelection.java

```diff
@@ -0,0 +1,246 @@
+package edu.umw.cpsc.collegediversity;
+
+import java.util.*;
+import sim.util.*;
+import ec.util.*;
+
+
+/**
+ * A Singleton class that can {@link #assign} a set of {@link
+ * Student}s to their non-freshmen dorms. The students in question are all
+ * those <i>except</i> incoming freshmen. 
+ * @see FreshmanHousingSelection
+ */
+public class UpperclassHousingSelection{
+
+    private Bag dorms;
+    private static UpperclassHousingSelection theInstance;
+
+    private UpperclassHousingSelection(){
+
+        dorms = new Bag();
+
+        //actually mixed-year but for simplicity I'm keeping it 
+        // upperclassman only for now
+        Dorm Arrington = new Dorm("Arrington", false, false, 74);
+
+        Dorm Ball = new Dorm("Ball", true, false, 53);
+        Dorm Custis = new Dorm("Custis", false, false, 21);
+        Dorm EagleLanding = new Dorm ("Eagle Landing", false, false, 312);
+        Dorm Framar = new Dorm("Framar", false, false, 11);
+        Dorm Madison = new Dorm("Madison", false, false, 21);
+        Dorm Marshall = new Dorm("Marshall", false, false, 74); 
+        Dorm Mason = new Dorm("Mason", false, false, 93); 
+        Dorm Westmoreland = new Dorm("Westmoreland", false, false, 56); 
+
+        //Willard houses students in single rooms which is not accounted for yet
+        Dorm Willard = new Dorm("Willard", false, false, 91); 
+
+        //100 of the rooms in the aparments are singles which is not 
+        // accounted for yet
+        Dorm Apartments = new Dorm("Apartments", false, false, 381); 
+
+        //for extraneous students who commute or live off campus
+        Dorm Offcampus = new Dorm ("Offcampus", false, false, 3500); 
+
+        dorms.add(Ball);
+        dorms.add(Arrington);
+        dorms.add(Custis);
+        dorms.add(Framar);
+        dorms.add(Madison);
+        dorms.add(Mason);
+        dorms.add(Westmoreland);
+        dorms.add(Willard);
+        dorms.add(Apartments);
+        dorms.add(EagleLanding);
+        dorms.add(Offcampus);
+    }
+
+    /**
+     * Singleton pattern.
+     */
+    public static UpperclassHousingSelection instance(){
+        if(theInstance==null){
+            theInstance=new UpperclassHousingSelection();
+        }
+        return theInstance;
+    }
+
+
+    /** 
+     * Assign the set of students (presumably non-incoming-freshmen, but
+     * this is not checked) passed to their dorm rooms, in a way that does 
+     * not take race into account. (compare {@link #assignByRace}.) As a 
+     * precursor (side effect) to this, any students already existing in 
+     * upperclass dorms will be removed.
+     */
+    public void assign(Bag students){
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 16. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.